### PR TITLE
Improvement: Do not run map on a single document in MapReduceDocumentsChain

### LINKED ISSUE #1937: Improvement: MapReduce summarization chains executes a map step on a single document
When feeding the map_reduce summarization chain with a single document, the doc is run through an unnecessary map step before running a combine prompt on it. 
The combine prompt would imo be sufficient to avoid summarizing a single summary which then is very lossy. Great project, hope this thrives!

### PR DESCRIPTION
**Pull request by TomTom101**
If a single document is provided, the mapping step in `acombine_docs()` will be skipped. Before, the document was summarized using the map_prompt and then again using the combine_prompt. Assumption is, that this single document fits into the context window.

Replaced self._process_results() with an async version and call with asyncio.run() from the sync methods

Fixes #1937 

#### Who can review?
@hwchase17 

Tag maintainers/contributors who might be interested:
@agola11

### CODE REVIEW

File: langchain/chains/combine_documents/map_reduce.py [hunk 1]
```diff
@@ -141,31 +141,62 @@ def combine_docs(
...
+
+        try:
+            import nest_asyncio
+
+            nest_asyncio.apply()
+            return asyncio.run(
+                self._aprocess_results(results, docs, callbacks=callbacks, **kwargs)
+            )
+        except ImportError:
+            raise ImportError(
```
Comment (hwchase17) [inline langchain/chains/combine_documents/map_reduce.py:164]: this is really common operation - we cant have this depend on optional imports. can we make this change in a separate pr? isnt related to title of PR of "do not run map on a single document"

Comment (TomTom101) [reply langchain/chains/combine_documents/map_reduce.py:164]: Thanks for taking the time!

True that, makes sense. I am trying to avoid code duplication of sync/async methods like _(a)process_results() but I guess there is no way around having two versions of them (like so often in the codebase) and possibly refactor them a little to minimize duplication. NOT having an async `_aprocess_results()` imo is not acceptable: Calling `await chain.arun()` sort of promises that things are running concurrently which would not be the case with a sync only method.

Comment (TomTom101) [reply langchain/chains/combine_documents/map_reduce.py]: I provided a version without the nest_asyncio dependency.

**Comment by TomTom101** [comment]
Removed all changes that are strictly unrelated to the PR title. Will suggest other changes in a different PR.

Added unit test

**Comment by baskaryan** [comment]
hm are we sure that in a one-document scenario you wouldn't still want to run both map and reduce? they could each perform arbitrary functionality, making it difficult to know in the abstract if map can be skipped. e.g. map might extract specific info into a bulleted list and reduce might just remove redundant bullets. in which case you'd probably still want to run map (and could maybe even skip reduce) on a single doc?

**Comment by TomTom101** [comment]
@baskaryan You are absolutely right, I indeed was in the tunnel of summarization, ignoring the other use cases of map-reduce! I guess, the mapReduce strategy is just a bad a idea when dealing with a single doc that fits the context window.

I'll close this!

Thanks!
